### PR TITLE
Proof generation, oracle removal and relative threshold

### PR DESCRIPTION
**Pull request by martinkersner**
# Description

This PR updates the 1) proof generation, and allows to 2) remove oracle and 3) specify relative signature threshold.

## Proof generation

Every node (off-chain oracle) is defined by an index obtained during `addOracle` function execution. This index will stay with node until the node keeps extending its expiration time, or is removed (`removeOracle`) by contract owner. The node index is given to node operator and can be also fetched from `oracles` public array. Submitted proofs have to be sorted based on given indexes, otherwise the verification process fails.

## Oracle removal

Previously, we assumed that every node will be allowed to produce proofs until it keeps extending its expiration time, however, sometimes, it might be necessary to remove a certain node operator. New function `removeOracle` that accomplishes this functionality has been added.

## Relative threshold

Previously, threshold proof was defined as an minimum number of valid proofs. This PR integrates a relative definition of threshold proof (e.g. 50 %) and defines a default threshold that is used if there is no custom threshold setup.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.

## Summary by CodeRabbit

- **New Features**
	- Added constants `MIN_THRESHOLD` and `MAX_THRESHOLD`.
	- Introduced `threshold` variable and `oracles` array.
	- Enhanced proof verification logic with indexes and quorum calculation.
	- Added `splitBytesToChunks` function to handle proof splitting.
	- Implemented `quorum` function to calculate the quorum for a threshold.

- **Bug Fixes**
	- Adjusted access control modifiers and error messages in `Feed.sol`.
	- Improved error handling for out-of-bounds and non-ascending index submissions during proof verification.

- **Documentation**
	- Streamlined function declaration in the `IFeedProxy` interface for clarity.

- **Tests**
	- Expanded testing suite to cover new functionalities like oracle management, proof submission, and threshold settings.

### CODE REVIEW

File: contracts/v0.2/test/SubmissionProxy.t.sol [hunk 1]
```diff
@@ -36,38 +36,75 @@ contract SubmissionProxyTest is Test {
...
+        // add
+        submissionProxy.addOracle(oracle_);
+        assertEq(submissionProxy.oracles(0), oracle_);
+        assertGe(submissionProxy.whitelist(oracle_), block.timestamp);
+
+        // remove
+        submissionProxy.removeOracle(oracle_);
+        vm.expectRevert(bytes("")); // "EvmError: Revert"
+        submissionProxy.oracles(0);
+        assertEq(submissionProxy.whitelist(oracle_), block.timestamp);
```
[the remark this answers is not in the document]
Comment (martinkersner) [reply contracts/v0.2/test/SubmissionProxy.t.sol:51]: We do not reset timestamp to 0, but rather set it to the current `block.timestamp` which achieves the same result while keeping the specific address inaccessible in the future.

**Comment by nick-bisonai** [comment]
Sorry, I'm not convinced why duplicate signature check is required, can you give me more explanation?

File: contracts/v0.2/src/SubmissionProxy.sol [hunk 2]
```diff
@@ -20,15 +20,22 @@ contract SubmissionProxy is Ownable {
...
     uint256 public constant MAX_EXPIRATION = 365 days;
+    uint8 public constant MIN_THRESHOLD = 1;
+    uint8 public constant MAX_THRESHOLD = 100;
 
     uint256 public maxSubmission = 50;
     uint256 public expirationPeriod = 5 weeks;
+    uint8 public threshold = 50; // 50 %
+    address[] public oracles;
+
     mapping(address oracle => uint256 expirationTime) public whitelist;
```
Comment (nick-bisonai) [inline contracts/v0.2/src/SubmissionProxy.sol:31]: How about keeping the index mapping from the contract and not taking from submit parameter?

```suggestion
   struct oracleInfo {
   index uint8;
   expirationTime uint256;
   }

    mapping(address oracle => oracleInfo info) public whitelist;
```

Comment (martinkersner) [reply contracts/v0.2/src/SubmissionProxy.sol]: Updated at https://github.com/Bisonai/orakl/pull/1379/commits/a54e11ddccc1d3e2fec580ba0db18bdc3c0b699f

**Review by nick-bisonai — COMMENTED** [review]
lgtm!

**Pull request merged**